src/proxyp/Message.cc

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "proxyp/Elements.h"
+#include "proxyp/Message.h"
+#include "sbuf/Stream.h"
+#include "sbuf/StringConvert.h"
+#include "SquidConfig.h"
+#include "StrList.h"
+
+ProxyProtocol::Message::Message(const char *ver, const uint8_t cmd):
+    version_(ver),
+    command_(Two::Command(cmd)),
+    ignoreAddresses_(false)
+{}
+
+SBuf
+ProxyProtocol::Message::toMime() const
+{
+    SBufStream result;
+    for (const auto &p: PseudoHeaderFields)
+        result << p.first << ": " << getValues(p.second) << "\r\n";
+    // cannot reuse Message::getValues(): need the original TLVs layout
+    for (const auto &tlv: tlvs)
+        result << tlv.type << ": " << tlv.value << "\r\n";
+    return result.buf();
+}
+
+SBuf
+ProxyProtocol::Message::getValues(const uint32_t headerType, const char sep) const
+{
+    SBufStream result;
+    char ipBuf[MAX_IPSTRLEN];
```

Comment:
Please move to the `if` statements that need it. Most do not so it makes sense to localize this ugliness.